### PR TITLE
Some cleanup in base

### PR DESCRIPTION
**Pull request by hbina**
While reading through the codebase and learning about Vulkan, I noticed the following changes in base/

Edit: This PR is turning out to be more than just using STL.

Edit 2: Comprehensive list from my second commit:

```
1. Some member functions can be made `const`.
2. Avoid using C-style casting.
3. explicit constructors for constructor with 1 argument.
4. Use STL algorithms whenever possible. Also pre-allocate buffer before pushing.
5. Removed unreachable code (like `break` after `return` and `exit` after `return`).
6. Initialized many previously uninitialized variables.
7. Handle error return codes (mostly from freopen_s in Windows).
8. Avoid flushing with every print statement by replacing `std::endl` with "\n".
9. Some off-by-one error involving finding a random ratio algorithms. Mostly just avoid the warning by explicitly casting RAND_MAX to float.
```

### CODE REVIEW

**Comment by SaschaWillems** [comment]
I haven't had time to take a look at the PR yet, but could you please either disable travis CI on your fork or change the E-Mail? I'm getting notifications for all your failed builds.

**Comment by SaschaWillems** [comment]
It would also be good if you could split up your PRs into smaller pieces. Some of the changes are okay, but other I'd like to decline. E.g. using std::find. I prefer raw loops due to better readability.

File: base/benchmark.hpp [hunk 1]
```diff
@@ -36,8 +36,12 @@ namespace vks
...
 #if defined(_WIN32)
 			AttachConsole(ATTACH_PARENT_PROCESS);
-			freopen_s(&stream, "CONOUT$", "w+", stdout);
-			freopen_s(&stream, "CONOUT$", "w+", stderr);
+			if (freopen_s(&stream, "CONOUT$", "w+", stdout)) {
+				std::cerr << "Unable to redirect stdout to file" << "\n";
+			}
+			if (freopen_s(&stream, "CONOUT$", "w+", stderr)) {
+				std::cerr << "Unable to redirect stdout to file" << "\n";
+			}
```
Comment (eloj) [inline base/benchmark.hpp:44]: "... stderr to file"

**Comment by hbina** [comment]
Oops, sorry about that. I will get back on it after I did my homework. I have since turned this PR into a Draft. I hope it doesn't trigger the emails.

**Pull request closed without merging**